File: src/main/java/com/larryhsiao/clotho/dgist/HmacSHA1.java
package com.larryhsiao.clotho.dgist;

import com.larryhsiao.clotho.Source;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import static java.nio.charset.StandardCharsets.UTF_8;

/**
 * HmacSHA1 with given data and secret.
 */
public class HmacSHA1 implements Source<byte[]> {
    private final Source<byte[]> data;
    private final String secret;

    public HmacSHA1(Source<byte[]> data, String secret) {
        this.data = data;
        this.secret = secret;
    }

    @Override
    public byte[] value() {
        try {
            final SecretKeySpec signingKey = new SecretKeySpec(
                secret.getBytes(UTF_8),
                "HmacSHA1"
            );
            final Mac mac = Mac.getInstance("HmacSHA1");
            mac.init(signingKey);
            return mac.doFinal(data.value());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/com/silverhetch/clotho/encryption/HmacSHA1.kt
package com.silverhetch.clotho.encryption

import com.silverhetch.clotho.Source
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * HmacSHA1 Encode with given data and secret.
 */
class HmacSHA1(private val data: Source<ByteArray>, private val secret: String) : Source<ByteArray> {
    override fun value(): ByteArray {
        val signingKey = SecretKeySpec(secret.toByteArray(Charsets.UTF_8), "HmacSHA1")
        val mac = Mac.getInstance("HmacSHA1")
        mac.init(signingKey)
        return mac.doFinal(data.value())
    }
}